plugins/inputs/statsd/datadog.go

```diff
@@ -0,0 +1,173 @@
+	}
+
+	rawTitle := message[:titleLen]
+	rawText := message[titleLen+1 : titleLen+1+textLen]
+	message = message[titleLen+1+textLen:]
+
+	if len(rawTitle) == 0 || len(rawText) == 0 {
+		return fmt.Errorf("Invalid event message format: empty 'title' or 'text' field")
+	}
+
+	// Handle hostname, with a priority to the h: field, then the host:
+	// tag and finally the defaultHostname value
+	// Metadata
+	m := cachedEvent{
+		name: rawTitle,
+	}
+	m.tags = make(map[string]string, strings.Count(message, ",")+2) // allocate for the approximate number of tags
+	m.fields = make(map[string]interface{}, 9)
+	m.fields["alert-type"] = "info" // default event type
+	m.fields["text"] = uncommenter.Replace(string(rawText))
+	m.tags["source"] = defaultHostname
+	m.fields["priority"] = priorityNormal
+	m.ts = now
+	if len(message) == 0 {
+		s.events = append(s.events, m)
+		return nil
+	}
+
+	if len(message) > 1 {
+		rawMetadataFields := strings.Split(message[1:], "|")
+		for i := range rawMetadataFields {
+			if len(rawMetadataFields[i]) < 2 {
+				log.Printf("W! [inputs.statsd] too short metadata field")
+			}
+			switch rawMetadataFields[i][:2] {
+			case "d:":
+				ts, err := strconv.ParseInt(rawMetadataFields[i][2:], 10, 64)
+				if err != nil {
+					log.Printf("W! [inputs.statsd] skipping timestamp: %s", err)
+					continue
+				}
+				m.fields["ts"] = ts
+			case "p:":
+				switch rawMetadataFields[i][2:] {
+				case priorityLow:
+					m.fields["priority"] = priorityLow
+				case priorityNormal: // we already used this as a default
+				default:
+					log.Printf("W! [inputs.statsd] skipping priority")
+					continue
+				}
+			case "h:":
+				m.tags["source"] = rawMetadataFields[i][2:]
+			case "t:":
+				switch rawMetadataFields[i][2:] {
+				case "error":
+					m.fields["alert-type"] = "error"
+				case "warning":
+					m.fields["alert-type"] = "warning"
+				case "success":
+					m.fields["alert-type"] = "success"
+				case "info": // already set for info
+				default:
+					log.Printf("W! [inputs.statsd] skipping alert type")
+					continue
+				}
+			case "k:":
+				// TODO(docmerlin): does this make sense?
+				m.tags["aggregation-key"] = rawMetadataFields[i][2:]
+			case "s:":
+				m.fields["source-type-name"] = rawMetadataFields[i][2:]
+			default:
+				if rawMetadataFields[i][0] == '#' {
+					parseDataDogTags(m.tags, rawMetadataFields[i][1:])
+				} else {
+					log.Printf("W! [inputs.statsd] unknown metadata type: '%s'", rawMetadataFields[i])
+				}
+			}
+		}
+	}
+	// host is a magic tag in the system, and it expects it to replace the result of h: if it is present
+	// telegraf will add a"host" tag anyway with different meaning than dogstatsd, so we need to switch these out
+	if host, ok := m.tags["host"]; ok {
+		delete(m.tags, "host")
+		m.tags["source"] = host
+	}
+	s.Lock()
+	s.events = append(s.events, m)
+	s.Unlock()
+	return nil
+}
+
+func parseDataDogTags(tags map[string]string, message string) {
+	start, i := 0, 0
+	var k string
```

Comment:
@danielnelson because the tag key names are being dynamically created here, should they be converted to snake case?  Do the output plugins do this ?

Reply:
As far as I know, the outputs do not do this